src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerSessionMonitor.cs

```diff
@@ -186,113 +183,102 @@ private bool RemoveSession(int sessionId, out ProfilerSession session)
+        /// </summary>
+        private void StartStream(int id, ProfilerSession session)
         {
-            var events = new List<ProfilerEvent>();
-            try
-            {
-                if (session == null || session.XEventSession == null)
+            if(session.XEventSession != null  && session.XEventSession.Session != null && session.XEventSession.ConnectionDetails != null){
+                CancellationTokenSource threadCancellationToken = new CancellationTokenSource();
+                var connectionString = ConnectionService.BuildConnectionString(session.XEventSession.ConnectionDetails);
+                var eventStreamer = new XELiveEventStreamer(connectionString, session.XEventSession.Session.Name);
+                // Start streaming task here, will run until cancellation or error with the feed.
+                var task = eventStreamer.ReadEventStream(xEvent => HandleXEvent(xEvent, session), threadCancellationToken.Token);
+
+                task.ContinueWith(t =>
                 {
-                    return events;
-                }
-
-                var targetXml = session.XEventSession.GetTargetXml();
-
-                XmlDocument xmlDoc = new XmlDocument();
-                xmlDoc.LoadXml(targetXml);
-
-                var nodes = xmlDoc.DocumentElement.GetElementsByTagName("event");
-                foreach (XmlNode node in nodes)
-                {
-                    var profilerEvent = ParseProfilerEvent(node);
-                    if (profilerEvent != null)
+                    //If cancellation token is missing, that means stream was stopped by the client, do not notify in this case.
+                    CancellationTokenSource targetToken;
+                    if (monitoredCancellationTokenSources.TryGetValue(id, out targetToken))
                     {
-                        events.Add(profilerEvent);
+                        StopSession(session.XEventSession.Id);
                     }
-                }
+                }, TaskContinuationOptions.OnlyOnFaulted);
+
+                this.monitoredCancellationTokenSources.Add(id, threadCancellationToken);
+                session.IsStreaming = true;
             }
-            catch (XEventException)
-            {
-                SendStoppedSessionInfoToListeners(session.XEventSession.Id);
+            else {
                 ProfilerSession tempSession;
-                RemoveSession(session.XEventSession.Id, out tempSession);
-            }
-            catch (Exception ex)
-            {
-                Logger.Write(TraceEventType.Warning, "Failed to poll session. error: " + ex.Message);
-            }
-            finally
-            {
-                session.IsPolling = false;
+                RemoveSession(id, out tempSession);
+                throw new Exception(SR.SessionMissingDetails(id));
```

Comment:
I think a better error message here would be something along the lines of `Unable to start streaming session {0} due to missing session details`. 

Although I'm curious when you expect this to ever happen in the first place.

Reply:
It shouldn't as long as the profilersession object is created properly in previous steps